### Data Cleaning

#### Importing Packages and Loading Data

In [1]:
import numpy as np
import pandas as pd
from pathlib import Path

In [2]:
# Setting up Project Paths
project_root = Path.cwd().parent
raw_dir = project_root/"data"/"raw"
processed_dir = project_root/"data"/"processed"

In [3]:
# Loading Data Frame
frames = []
for f in sorted(raw_dir.rglob("*.csv")):
    df_part = pd.read_csv(f)
    df_part["source_file"] = f.name
    frames.append(df_part)

df = pd.concat(frames, ignore_index=True)

df

,Crime ID,Month,Reported by,Falls within,Longitude,Latitude,Location,LSOA code,LSOA name,Crime type,Last outcome category,Context,source_file
0,e7b720d0e1302d2d06db7b28b29132eb194864d44d7921...,2024-01,City of London Police,City of London Police,-0.106220,51.518275,On or near B500,E01000916,Camden 027B,Theft from the person,Investigation complete; no suspect identified,NaN,2024-01-city-of-london-street.csv
1,e60a5ac62a80e866453254474137c3206417422c62f0c0...,2024-01,City of London Police,City of London Police,-0.107682,51.517786,On or near B521,E01000917,Camden 027C,Other theft,Investigation complete; no suspect identified,NaN,2024-01-city-of-london-street.csv
2,986f618142ec52b7f254e4b0549da2f17ceeb0e130db6c...,2024-01,City of London Police,City of London Police,-0.111596,51.518281,On or near Chancery Lane,E01000914,Camden 028B,Other theft,Investigation complete; no suspect identified,NaN,2024-01-city-of-london-street.csv
3,05dc27a88748356f6d59b0bd1389710ebfb42b37e565af...,2024-01,City of London Police,City of London Police,-0.111596,51.518281,On or near Chancery Lane,E01000914,Camden 028B,Theft from the person,Status update unavailable,NaN,2024-01-city-of-london-street.csv
4,373d78e2ccec5d05a547cd4bee19045a9e050042a0e6e7...,2024-01,City of London Police,City of London Police,-0.112096,51.515942,On or near Nightclub,E01000914,Camden 028B,Theft from the person,Investigation complete; no suspect identified,NaN,2024-01-city-of-london-street.csv
...,...,...,...,...,...,...,...,...,...,...,...,...,...
2298607,d7bcaa710e61011a6192bd504e86738b100f47b42be464...,2025-12,Metropolitan Police Service,Metropolitan Police Service,-2.171098,52.568065,On or near Warstones Drive,E01010490,Wolverhampton 028B,Violence and sexual offences,Under investigation,NaN,2025-12-metropolitan-street.csv
2298608,60acb42bd7459bb8963f3009bdaf77e8246a1b525a3ce6...,2025-12,Metropolitan Police Service,Metropolitan Police Service,-1.851453,52.110001,On or near Little Road,E01032415,Wychavon 011D,Other crime,Under investigation,NaN,2025-12-metropolitan-street.csv
2298609,1d6177d0d7e23c56d48ea7511406118f0daa205d35c8b7...,2025-12,Metropolitan Police Service,Metropolitan Police Service,-1.870244,52.040086,On or near Station Road,E01032351,Wychavon 018A,Theft from the person,Investigation complete; no suspect identified,NaN,2025-12-metropolitan-street.csv
2298610,86578e92566319f2b8822dd2301965f40a9dfa7e9e6f43...,2025-12,Metropolitan Police Service,Metropolitan Police Service,-2.262334,52.389968,On or near Church Walk,E01032454,Wyre Forest 006D,Violence and sexual offences,Under investigation,NaN,2025-12-metropolitan-street.csv


#### Dropping Columns and Rows

In [4]:
# "Context" is fully empty
df = df.drop(columns=["Context"])

In [5]:
df.head()

,Crime ID,Month,Reported by,Falls within,Longitude,Latitude,Location,LSOA code,LSOA name,Crime type,Last outcome category,source_file
0,e7b720d0e1302d2d06db7b28b29132eb194864d44d7921...,2024-01,City of London Police,City of London Police,-0.106220,51.518275,On or near B500,E01000916,Camden 027B,Theft from the person,Investigation complete; no suspect identified,2024-01-city-of-london-street.csv
1,e60a5ac62a80e866453254474137c3206417422c62f0c0...,2024-01,City of London Police,City of London Police,-0.107682,51.517786,On or near B521,E01000917,Camden 027C,Other theft,Investigation complete; no suspect identified,2024-01-city-of-london-street.csv
2,986f618142ec52b7f254e4b0549da2f17ceeb0e130db6c...,2024-01,City of London Police,City of London Police,-0.111596,51.518281,On or near Chancery Lane,E01000914,Camden 028B,Other theft,Investigation complete; no suspect identified,2024-01-city-of-london-street.csv
3,05dc27a88748356f6d59b0bd1389710ebfb42b37e565af...,2024-01,City of London Police,City of London Police,-0.111596,51.518281,On or near Chancery Lane,E01000914,Camden 028B,Theft from the person,Status update unavailable,2024-01-city-of-london-street.csv
4,373d78e2ccec5d05a547cd4bee19045a9e050042a0e6e7...,2024-01,City of London Police,City of London Police,-0.112096,51.515942,On or near Nightclub,E01000914,Camden 028B,Theft from the person,Investigation complete; no suspect identified,2024-01-city-of-london-street.csv


In [6]:
df.shape

(2298612, 12)

In [7]:
# Dropping rows with no Locations
df = df.dropna(subset=["LSOA code", "Longitude", "Latitude"])
df.shape

(2296350, 12)

#### Changing Column Names

In [8]:
df.columns = (
    df.columns
    .str.strip()
    .str.lower()
    .str.replace(" ", "_")
)

df.head()

,crime_id,month,reported_by,falls_within,longitude,latitude,location,lsoa_code,lsoa_name,crime_type,last_outcome_category,source_file
0,e7b720d0e1302d2d06db7b28b29132eb194864d44d7921...,2024-01,City of London Police,City of London Police,-0.106220,51.518275,On or near B500,E01000916,Camden 027B,Theft from the person,Investigation complete; no suspect identified,2024-01-city-of-london-street.csv
1,e60a5ac62a80e866453254474137c3206417422c62f0c0...,2024-01,City of London Police,City of London Police,-0.107682,51.517786,On or near B521,E01000917,Camden 027C,Other theft,Investigation complete; no suspect identified,2024-01-city-of-london-street.csv
2,986f618142ec52b7f254e4b0549da2f17ceeb0e130db6c...,2024-01,City of London Police,City of London Police,-0.111596,51.518281,On or near Chancery Lane,E01000914,Camden 028B,Other theft,Investigation complete; no suspect identified,2024-01-city-of-london-street.csv
3,05dc27a88748356f6d59b0bd1389710ebfb42b37e565af...,2024-01,City of London Police,City of London Police,-0.111596,51.518281,On or near Chancery Lane,E01000914,Camden 028B,Theft from the person,Status update unavailable,2024-01-city-of-london-street.csv
4,373d78e2ccec5d05a547cd4bee19045a9e050042a0e6e7...,2024-01,City of London Police,City of London Police,-0.112096,51.515942,On or near Nightclub,E01000914,Camden 028B,Theft from the person,Investigation complete; no suspect identified,2024-01-city-of-london-street.csv


#### Removing out of London crimes

In [9]:
# The Metroploitan Police sometimes reports crimes which happen just outside the London boundary. Greater London sits roughly within longitude −0.52 to 0.33 and latitude 51.28 to 51.70. So we keep everything inside this boundary.

lon_min, lon_max = -0.52, 0.33
lat_min, lat_max = 51.28, 51.70

in_london = (
    df["longitude"].between(lon_min, lon_max)
    & df["latitude"].between(lat_min, lat_max)
)
df.shape

(2296350, 12)

In [10]:
df = df[in_london]
df.shape

(2290057, 12)

#### Duplicate Rows

In [11]:
df.duplicated().value_counts()

False    2073430
True      216627
Name: count, dtype: int64

In [12]:
df[df.duplicated()]

,crime_id,month,reported_by,falls_within,longitude,latitude,location,lsoa_code,lsoa_name,crime_type,last_outcome_category,source_file
268,NaN,2024-01,City of London Police,City of London Police,-0.078755,51.518440,On or near Artillery Lane,E01032739,City of London 001F,Anti-social behaviour,NaN,2024-01-city-of-london-street.csv
269,NaN,2024-01,City of London Police,City of London Police,-0.081113,51.516869,On or near Bishopsgate,E01032739,City of London 001F,Anti-social behaviour,NaN,2024-01-city-of-london-street.csv
525,678ea9f5a5067acde0a8101c7b43012e32959f79c789ca...,2024-01,City of London Police,City of London Police,-0.091540,51.511896,On or near Cannon Street,E01032739,City of London 001F,Violence and sexual offences,Status update unavailable,2024-01-city-of-london-street.csv
767,NaN,2024-01,Metropolitan Police Service,Metropolitan Police Service,0.129358,51.583686,On or near Gregory Road,E01000028,Barking and Dagenham 001B,Anti-social behaviour,NaN,2024-01-metropolitan-street.csv
768,NaN,2024-01,Metropolitan Police Service,Metropolitan Police Service,0.129358,51.583686,On or near Gregory Road,E01000028,Barking and Dagenham 001B,Anti-social behaviour,NaN,2024-01-metropolitan-street.csv
...,...,...,...,...,...,...,...,...,...,...,...,...
2298569,NaN,2025-12,Metropolitan Police Service,Metropolitan Police Service,-0.137230,51.486309,On or near Parking Area,E01035722,Westminster 024G,Anti-social behaviour,NaN,2025-12-metropolitan-street.csv
2298570,NaN,2025-12,Metropolitan Police Service,Metropolitan Police Service,-0.137230,51.486309,On or near Parking Area,E01035722,Westminster 024G,Anti-social behaviour,NaN,2025-12-metropolitan-street.csv
2298571,NaN,2025-12,Metropolitan Police Service,Metropolitan Police Service,-0.137230,51.486309,On or near Parking Area,E01035722,Westminster 024G,Anti-social behaviour,NaN,2025-12-metropolitan-street.csv
2298572,NaN,2025-12,Metropolitan Police Service,Metropolitan Police Service,-0.137230,51.486309,On or near Parking Area,E01035722,Westminster 024G,Anti-social behaviour,NaN,2025-12-metropolitan-street.csv


In [16]:
dup_mask = df.duplicated(keep="first")

In [18]:
df.loc[dup_mask, "crime_type"].value_counts()

crime_type
Anti-social behaviour           214243
Violence and sexual offences      1109
Public order                       685
Drugs                              323
Robbery                             95
Criminal damage and arson           48
Burglary                            45
Theft from the person               22
Vehicle crime                       19
Other theft                         17
Possession of weapons               10
Other crime                          5
Shoplifting                          5
Bicycle theft                        1
Name: count, dtype: int64

##### Anti-social behavior has no crime ID and has only month level dates. Separate incidents of the same kind are being flagged as duplicate but are real separate incidents and we shall keep them in our data. We will remove the duplicates that have a crime ID.

In [25]:
has_cid = df["crime_id"].notna()
is_dup = df.duplicated(keep="first")

df.shape

(2290057, 12)

In [26]:
dropping = has_cid & is_dup
df = df[~dropping]

df.shape

(2287673, 12)

In [27]:
df

,crime_id,month,reported_by,falls_within,longitude,latitude,location,lsoa_code,lsoa_name,crime_type,last_outcome_category,source_file
0,e7b720d0e1302d2d06db7b28b29132eb194864d44d7921...,2024-01,City of London Police,City of London Police,-0.106220,51.518275,On or near B500,E01000916,Camden 027B,Theft from the person,Investigation complete; no suspect identified,2024-01-city-of-london-street.csv
1,e60a5ac62a80e866453254474137c3206417422c62f0c0...,2024-01,City of London Police,City of London Police,-0.107682,51.517786,On or near B521,E01000917,Camden 027C,Other theft,Investigation complete; no suspect identified,2024-01-city-of-london-street.csv
2,986f618142ec52b7f254e4b0549da2f17ceeb0e130db6c...,2024-01,City of London Police,City of London Police,-0.111596,51.518281,On or near Chancery Lane,E01000914,Camden 028B,Other theft,Investigation complete; no suspect identified,2024-01-city-of-london-street.csv
3,05dc27a88748356f6d59b0bd1389710ebfb42b37e565af...,2024-01,City of London Police,City of London Police,-0.111596,51.518281,On or near Chancery Lane,E01000914,Camden 028B,Theft from the person,Status update unavailable,2024-01-city-of-london-street.csv
4,373d78e2ccec5d05a547cd4bee19045a9e050042a0e6e7...,2024-01,City of London Police,City of London Police,-0.112096,51.515942,On or near Nightclub,E01000914,Camden 028B,Theft from the person,Investigation complete; no suspect identified,2024-01-city-of-london-street.csv
...,...,...,...,...,...,...,...,...,...,...,...,...
2298579,8cfa73aac9ffd2e8f0d597b4b1e529bacc2cc3e82d509a...,2025-12,Metropolitan Police Service,Metropolitan Police Service,-0.137230,51.486309,On or near Parking Area,E01035722,Westminster 024G,Violence and sexual offences,Under investigation,2025-12-metropolitan-street.csv
2298580,4822666edb3d22a5680ec316ce7c5cef0d861d659b3fa2...,2025-12,Metropolitan Police Service,Metropolitan Police Service,-0.135320,51.487600,On or near St George'S Square,E01035722,Westminster 024G,Violence and sexual offences,Unable to prosecute suspect,2025-12-metropolitan-street.csv
2298581,e86e036bb9da06e7cdfa8ffc5caa4d3e5bb8f68d21a24a...,2025-12,Metropolitan Police Service,Metropolitan Police Service,-0.133538,51.487500,On or near Aylesford Street,E01035722,Westminster 024G,Violence and sexual offences,Action to be taken by another organisation,2025-12-metropolitan-street.csv
2298582,f20a67354e3c9415f5e0281f9e084f1dbae4e858ad0b62...,2025-12,Metropolitan Police Service,Metropolitan Police Service,-0.133538,51.487500,On or near Aylesford Street,E01035722,Westminster 024G,Violence and sexual offences,Unable to prosecute suspect,2025-12-metropolitan-street.csv


In [28]:
# Dropping Falls Within column
df = df.drop(columns=["falls_within"])

df

,crime_id,month,reported_by,longitude,latitude,location,lsoa_code,lsoa_name,crime_type,last_outcome_category,source_file
0,e7b720d0e1302d2d06db7b28b29132eb194864d44d7921...,2024-01,City of London Police,-0.106220,51.518275,On or near B500,E01000916,Camden 027B,Theft from the person,Investigation complete; no suspect identified,2024-01-city-of-london-street.csv
1,e60a5ac62a80e866453254474137c3206417422c62f0c0...,2024-01,City of London Police,-0.107682,51.517786,On or near B521,E01000917,Camden 027C,Other theft,Investigation complete; no suspect identified,2024-01-city-of-london-street.csv
2,986f618142ec52b7f254e4b0549da2f17ceeb0e130db6c...,2024-01,City of London Police,-0.111596,51.518281,On or near Chancery Lane,E01000914,Camden 028B,Other theft,Investigation complete; no suspect identified,2024-01-city-of-london-street.csv
3,05dc27a88748356f6d59b0bd1389710ebfb42b37e565af...,2024-01,City of London Police,-0.111596,51.518281,On or near Chancery Lane,E01000914,Camden 028B,Theft from the person,Status update unavailable,2024-01-city-of-london-street.csv
4,373d78e2ccec5d05a547cd4bee19045a9e050042a0e6e7...,2024-01,City of London Police,-0.112096,51.515942,On or near Nightclub,E01000914,Camden 028B,Theft from the person,Investigation complete; no suspect identified,2024-01-city-of-london-street.csv
...,...,...,...,...,...,...,...,...,...,...,...
2298579,8cfa73aac9ffd2e8f0d597b4b1e529bacc2cc3e82d509a...,2025-12,Metropolitan Police Service,-0.137230,51.486309,On or near Parking Area,E01035722,Westminster 024G,Violence and sexual offences,Under investigation,2025-12-metropolitan-street.csv
2298580,4822666edb3d22a5680ec316ce7c5cef0d861d659b3fa2...,2025-12,Metropolitan Police Service,-0.135320,51.487600,On or near St George'S Square,E01035722,Westminster 024G,Violence and sexual offences,Unable to prosecute suspect,2025-12-metropolitan-street.csv
2298581,e86e036bb9da06e7cdfa8ffc5caa4d3e5bb8f68d21a24a...,2025-12,Metropolitan Police Service,-0.133538,51.487500,On or near Aylesford Street,E01035722,Westminster 024G,Violence and sexual offences,Action to be taken by another organisation,2025-12-metropolitan-street.csv
2298582,f20a67354e3c9415f5e0281f9e084f1dbae4e858ad0b62...,2025-12,Metropolitan Police Service,-0.133538,51.487500,On or near Aylesford Street,E01035722,Westminster 024G,Violence and sexual offences,Unable to prosecute suspect,2025-12-metropolitan-street.csv


#### Saving the Dataframe

In [31]:
print("Shape:", df.shape)
print("\nRemaining missing values:")
print(df.isna().sum())

Shape: (2287673, 11)

Remaining missing values:
crime_id                 467822
month                         0
reported_by                   0
longitude                     0
latitude                      0
location                      0
lsoa_code                     0
lsoa_name                     0
crime_type                    0
last_outcome_category    467822
source_file                   0
dtype: int64


In [ ]:
df.to_parquet(processed_dir/"london_crime.parquet", index=False)

#### Summary

| Step | Action | Rows |
|---|---|--|
| Start | Loaded 48 raw CSVs | 2,298,612 |
| 1 | Dropped `Context` (100% empty) | 2,298,612 |
| 2 | Dropped rows with no location | 2,296,350 |
| 3 | Standardised column names |  |
| 4 | Removed out-of-London crimes | 2,290,057 |
| 5 | Removed true duplicates (ID'd only) | 2,287,673 |
| 6 | Dropped redundant `falls_within` | 2,287,673 |
| **Final** | **Saved `london_crime.parquet`** | **2,287,673** |

98.9% of exact duplicates were Anti-social behaviour, which
has no Crime ID. These are legitimate distinct incidents made identical by
month-only dates and similar locations. Removing them would corrupt the
prediction target. Only the 2,384 duplicates with a Crime ID were removed.